docs/frontend-system/architecture/20-extensions.md

```diff
@@ -101,59 +95,55 @@ The `ExtensionDataRef` can then be used to describe an output property of the ex
+Note that you are **not** allowed to repeat the same data reference in the outputs, or return multiple values for the same reference. Multiple outputs for the same reference will conflict with each other and cause an error. If you want to output multiple values of the same TypeScript type you should create separate references for each value. That in turn means that overly generic extension data references a bad idea, for example a generic "string" type. Instead create separate references for each type of data that you want to share.
 
 ```tsx
 const extension = createExtension({
   // ...
-  output: {
-    // ❌ Bad example - outputting values of same type
-    element1: reactElementExtensionDataRef,
-    element2: reactElementExtensionDataRef,
-  },
+  output: [
+    // ❌ Bad example - duplicate output declaration
+    reactElementExtensionDataRef,
+    reactElementExtensionDataRef,
+  ],
   factory() {
-    return {
-      element1: <div>Hello World</div>,
-      element2: <div>Hello World</div>,
-    };
+    return [
+      // ❌ Bad example - duplicate output values
+      reactElementExtensionDataRef(<div>Hello</div>),
+      reactElementExtensionDataRef(<div>World</div>),
+    ];
   },
 });
 ```
 
-### Core Extension Data
+### Core extension data
 
 We provide default `coreExtensionData`, which provides commonly used `ExtensionDataRef`s - e.g. for `React.JSX.Element` and `RouteRef`. They can be used when creating your own extension. For example, the React Element extension data that we defined above is already provided as `coreExtensionData.reactElement`.
 
 For a full list and explanations of all types of core extension data, see the [core extension data reference](../building-plugins/04-built-in-data-refs.md).
 
-### Optional Extension Data
+### Optional extension data
 
 By default all extension data is required, meaning that the extension factory must provide a value for each output. However, it is possible to make extension data optional by calling the `.optional()` method. This makes it optional for the factory function to return a value as part of its output. When calling the `.optional()` method you create a new copy of the extension data reference, it does not mutate the existing reference.
 
 ```tsx
 const extension = createExtension({
   // ...
-  output: {
-    element: coreExtensionData.reactElement.optional(),
-  },
+  output: [coreExtensionData.reactElement.optional()],
   factory() {
-    return {
-      element:
+    return [
+      coreExtensionData.reactElement.optional()(
```

Comment:
Really? We shouldn't have to have optional() here too right? Does this example need to be rewritten as a generator or something using like .flat() or so?

Reply:
Ah yep to both of those.